draft-ietf-ppm-dap.md

```diff
@@ -602,12 +615,16 @@ differential privacy.
 For this query type, the Aggregators group reports into arbitrary batches such
 that each batch has roughly the same number of reports. These batches are
 identified by opaque "batch IDs", allocated in an arbitrary fashion by the
-Leader. To get the aggregate of a batch, the Collector issues a query specifying
-the batch ID of interest (see {{query}}).
+Leader.
+
+To get the aggregate of a batch, the Collector issues a query specifying the
+batch ID of interest (see {{query}}). The Collector may not know which batch ID
+it is interested in; in this case, it can also issue a query of type
+`current-batch`, which allows the Leader to select a recent batch to aggregate.
```

Comment:
We might add a SHOULD here to guide the Leader towards returning the oldest, not-yet-collected batch.

Reply:
I'm not sure this notion can be meaningful. The leader necessarily keeps track of the query count against a batch for privacy reasons, but unfortunately I don't think that means it can correctly maintain the position of the `current-batch` cursor. Suppose the leader services a collect request and delivers an aggregate result to the collector, but the collector crashes while receiving the message. The leader now might consider the batch in question collected and thus advance the cursor to the next batch. But when the collector restarts and queries `current-batch`, it really should get the batch that the leader just advanced past. Without an explicit ack from the collector, the leader has to assume the result has not been delivered.

The more I think about it, the more I think that this is the semantics of a message queue like AWS SQS or GCP PubSub, both of which will redeliver messages to consumers until a message is explicitly acked.

Anyway, the best the leader can do given what's in this PR is to advance the cursor regardless of the message being received by the collector. This means that some batches could be lost, but this proposal is still workable in the common case where the collector doesn't crash.

Reply:
I'm happy to add a SHOULD if folks think it wise.

FWIW, I did not intend to imply list-like semantics for `current-batch` -- I attempted to be careful not to imply anything about the ordering of batches returned by `current-batch`. A few plausible implementation strategies:

* A queue, as discussed upthread. (and yep, I think `current-batch` would allow a fairly nice implementation based on GCP Pub/Sub, AWS SQS, or similar technologies)
* Randomly choose an acceptable batch with each call to `current-batch`.
* For a sharded implementation: choose a random shard, and ask it for a current batch. (each shard might have its own queue, select randomly, or use some other strategy)

Thoughts? I purposefully left the implementation strategy ambiguous (on the idea that it's often wise to be less prescriptive rather than more prescriptive when it comes to implementation strategies), but if everyone thinks a queue is most appropriate then we can certainly suggest that. A queue or queue-like implementation is what I'd go with, but I didn't want to block implementations from experimenting with different selection methods.

===

> Anyway, the best the leader can do given what's in this PR is to advance the cursor regardless of the message being received by the collector. This means that some batches could be lost, but this proposal is still workable in the common case where the collector doesn't crash.

Can you elaborate? With `current-batch`, the batch can continue to be returned until the collector actually attempts collection for the first time. This means that an ill-timed crash can be recovered from by polling `current-batch` again -- the same batch will eventually be returned again. The expectation is that the collector will durably store the collect URL before beginning to poll.

Or, to put it another way:
* If the crash happens before the collector durably stores the collect URI, the collector will eventually receive the same batch back from `current-batch` again -- no data loss.
* If the crash happens after the collector durably stores the collect URI, the collector has the collect URI in its durable storage & can begin/continue polling -- no data loss.

Reply:
Fair points, both :) We probably SHOULDn't then.

Reply:
Looking again: I added text that the Leader `SHOULD` return a batch that has not yet began collection, as the expected semantics there were IMO not clear enough. I left off the ordering part for now, but I'm happy to add it if folks think we should enforce queue-like behavior.

Reply:
LGTM

Reply:
I'm wary of responding to a resolved thread in a PR that I approved already but I want to note one more thing in this context.

branlwyd:
>Can you elaborate? With current-batch, the batch can continue to be returned until the collector actually attempts collection for the first time. This means that an ill-timed crash can be recovered from by polling current-batch again -- the same batch will eventually be returned again. The expectation is that the collector will durably store the collect URL before beginning to poll.

IIUC, you're saying that the collector doing `POST /collect` with `current-batch` is akin to a subscriber pulling a message, and then the first `GET {collect-uri}` is akin to acking the message, in that it indicates to the leader that the collector has received the collect URI and thus the leader can then respond to `current-batch` with a different collect URI? I think that makes sense, and I see how this is fault tolerant, but now it seems like `GET {collect-uri}` has a side effect of changing how the server responds to `POST /collect`, which seems a bit off. So I wonder if we should change it to `POST {collect-uri}`. But I don't think we should make that change here, as we need to consider more broadly what abstraction we want for delivering aggregate shares from aggregators to collectors.

Reply:
> IIUC, you're saying that the collector doing POST /collect with current-batch is akin to a subscriber pulling a message, and then the first GET {collect-uri} is akin to acking the message, in that it indicates to the leader that the collector has received the collect URI and thus the leader can then respond to current-batch with a different collect URI?

Correct, though I would note that the Leader is free to return a different batch even before the Collector "acks" it by beginning collection. (This would be handy to support use-cases where the Collector wants to collect potentially many batches in parallel -- no need to synchronize the "receive a `current-batch` for collection" logic with the "poll a collect URI" logic, they can run more-or-less independently.) Or, to put it another way: "acking" a batch removes it from the set of batches that can be returned from `current-batch`; there is no requirement for `current-batch` to continuously return the same batch until it is "acked".


> I think that makes sense, and I see how this is fault tolerant, but now it seems like GET {collect-uri} has a side effect of changing how the server responds to POST /collect, which seems a bit off. So I wonder if we should change it to POST {collect-uri}. But I don't think we should make that change here, as we need to consider more broadly what abstraction we want for delivering aggregate shares from aggregators to collectors.

Oh, that's a good point -- this is indeed a state change via a `GET`. While I think it is technically-speaking safe/non-problematic in this particular instance, it's worthy of being cleaned up to follow best practices. The most direct way would be to switch to `POST`, as you suggest.